File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for(char a : s.toCharArray()) {
            if (a == ')') {
                if (stack.isEmpty() || stack.peek() != '(') {
                    return false;
                }
                stack.pop();
            } else if (a == ']') {
                if (stack.isEmpty() || stack.peek() != '[') {
                    return false;
                }
                stack.pop();

            } else if (a == '}') {
                if (stack.isEmpty() || stack.peek() != '{') {
                    return false;
                }
                stack.pop();

            } else {
                stack.push(a);
            }   
        }

        if (stack.isEmpty()) {
            return true;
        }

        return false;
        
        
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for char in s:
            if len(stack) == 0:
                stack.append(char)
                continue
                
            if char is ']':
                if stack[-1] != '[':
                    return False
                stack.pop()    
            elif char is '}':
                if stack[-1] != '{':
                    return False
                stack.pop()
            elif char is ')':
                if stack[-1] != '(':
                    return False        
                stack.pop()
            else:
                stack.append(char)    

        return len(stack) == 0        

